### PR TITLE
Improve stale pull request logic

### LINKED ISSUE #38: Post new warning if a commit is added then PR goes stale again
At the moment, if any warning about closing a PR has been posted to a PR, we never show the warning again. This poses an issue for the case where someone adds a commit and the PR goes stale again (no new warning is posted and the PR is closed 6 months later).

When searching for previous comments, we should restrict ourselves to comments that have been made since the last commit date. We should adjust the ``find_comments`` function to take a start time.

### LINKED ISSUE #89: Issue with invalid label
Seen in the log of the auto issue closer:

```
018-07-18T20:57:43.980874+00:00 app[web.1]: Checking 7060
2018-07-18T20:57:44.485682+00:00 app[web.1]: -> CLOSING issue 7060
2018-07-18T20:57:44.739104+00:00 app[web.1]: -> WARNING: Label does not exist: set()
2018-07-18T20:57:45.494535+00:00 app[web.1]: Checking 7054
2018-07-18T20:57:45.967184+00:00 app[web.1]: -> CLOSING issue 7054
2018-07-18T20:57:46.172075+00:00 app[web.1]: -> WARNING: Label does not exist: set()
2018-07-18T20:57:47.053923+00:00 app[web.1]: Checking 7053
2018-07-18T20:57:47.681967+00:00 app[web.1]: -> WARNING issue 7053
2018-07-18T20:57:47.946722+00:00 app[web.1]: Checking 7045
2018-07-18T20:57:48.370363+00:00 app[web.1]: -> CLOSING issue 7045
2018-07-18T20:57:48.573939+00:00 app[web.1]: -> WARNING: Label does not exist: set()
```

### PR DESCRIPTION
**Pull request by astrofrog**
This changes the logic of the stale PR blueprint in the following way

* Closing is only ever done if a warning has been posted before, regardless of the time since the last commit.

* The ``stale_pull_requests_close_seconds`` now has a different meaning - it is the time since the last warning after which a pull request is closed (not since the latest commit). This ensures that even if the bot is late in warning, the closing always happens a fixed time after the warning.

* Warnings about closing that pre-date the latest commit are ignored. Fixes https://github.com/astropy/astropy-bot/issues/38

* Fixed some of the logic about missing labels. Fixes https://github.com/astropy/astropy-bot/issues/89

* If the keep open label is present, no warning is emitted (somehow the test seemed to suggest a warning would be emitted previously)
 
Once this is merged, I will update the Heroku config to modify the default value of ``stale_pull_requests_close_seconds``.

### CODE REVIEW

File: changebot/blueprints/stale_pull_requests.py [hunk 1]
```diff
@@ -86,37 +86,51 @@ def process_pull_requests(repository, installation):
...
             continue
+
         commit_time = pr.last_commit_date
+        time_since_last_commit = now - commit_time
+
+        # Note: if warning time is before commit time, it's as if the warning
+        # didn't exist since it's no longer relevant.
+        warning_time = pr.last_comment_date('astropy-bot[bot]', filter_keep=is_close_warning)
+        if warning_time is None or warning_time < commit_time:
+            has_warning = False
```
Comment (pllim) [inline changebot/blueprints/stale_pull_requests.py]: Is the `has_warning` flag necessary? It seems like setting `time_since_last_warning = 0` has the same effect anyway, unless you allow `stale_pull_requests_close_seconds` to be negative, which I don't think you do.

File: changebot/blueprints/tests/test_stale_pull_requests.py [hunk 2]
```diff
@@ -172,30 +180,29 @@ def test_close_disabled(self):
...
             with patch.object(app, 'stale_pull_requests_close', False):
                 # The list() call is to forge the generator to run fully
                 list(process_pull_requests('repo', 'installation'))
 
-        assert self.submit_comment.call_count == 1
-        expected = PULL_REQUESTS_CLOSE_WARNING.format(pasttime='4 minutes', futuretime='20 seconds')
-        self.submit_comment.assert_called_with(expected)
-        assert self.close.call_count == 0
-        assert self.set_labels.call_count == 0
+        assert self.submit_comment.call_count == 0
```
Comment (pllim) [inline changebot/blueprints/tests/test_stale_pull_requests.py:196]: Yes, I think this is now correct. Not sure why the test expected a warning with the label.

**Review by pllim — APPROVED** [review]
This is API change that is not backward compatible. While I don't know how many out there who use this bot for their own projects, I am pinging @drdavella and @Cadair , just in case.

Otherwise LGTM!

File: changebot/blueprints/stale_pull_requests.py (hunk 1 above)
Comment (astrofrog) [reply changebot/blueprints/stale_pull_requests.py]: Good catch!

**Pull request merged**